src/java/org/apache/cassandra/index/sai/disk/v4/V4InvertedIndexSearcher.java

```diff
@@ -0,0 +1,94 @@
+import org.apache.cassandra.index.sai.disk.v1.InvertedIndexSearcher;
+import org.apache.cassandra.index.sai.disk.v1.PerIndexFiles;
+import org.apache.cassandra.index.sai.disk.v1.SegmentMetadata;
+import org.apache.cassandra.index.sai.metrics.MulticastQueryEventListeners;
+import org.apache.cassandra.index.sai.metrics.QueryEventListener;
+import org.apache.cassandra.index.sai.plan.Expression;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+
+public class V4InvertedIndexSearcher extends InvertedIndexSearcher
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    V4InvertedIndexSearcher(PrimaryKeyMap.Factory primaryKeyMapFactory,
+                            PerIndexFiles perIndexFiles,
+                            SegmentMetadata segmentMetadata,
+                            IndexDescriptor indexDescriptor,
+                            IndexContext indexContext) throws IOException
+    {
+        super(primaryKeyMapFactory, perIndexFiles, segmentMetadata, indexDescriptor, indexContext);
+    }
+
+    @Override
+    protected PostingList searchPosting(Expression exp, QueryContext context)
+    {
+        if (logger.isTraceEnabled())
+            logger.trace(indexContext.logMessage("Searching on expression '{}'..."), exp);
+
+        if (exp.getOp().isEquality() || exp.getOp() == Expression.Op.MATCH)
+        {
+            final ByteComparable term = V4OnDiskFormat.instance.encode(exp.lower.value.encoded, indexContext.getValidator());
+            QueryEventListener.TrieIndexEventListener listener = MulticastQueryEventListeners.of(context, perColumnEventListener);
+            return reader.exactMatch(term, listener, context);
+        }
+        else if (exp.getOp() == Expression.Op.RANGE)
+        {
+            var someType = indexContext.getValidator();
+            if (!(someType instanceof CompositeType))
+                throw new IllegalStateException("Only composite types are supported for range queries for now");
+            var type = (CompositeType) someType;
+
+            // TODO how do we make this generic so we don't explicitly reference the ByteBufferAccessor.instance here?
+            ByteComparable lower = null;
+            if (exp.lower != null)
+            {
+                var terminator = exp.lower.inclusive ? ByteSource.TERMINATOR : ByteSource.GT_NEXT_COMPONENT;
+                lower = v -> type.asComparableBytes(ByteBufferAccessor.instance, exp.lower.value.raw, v, terminator);
+            }
+            ByteComparable upper = null;
+            if (exp.upper != null)
+            {
+                var terminator = exp.upper.inclusive ? ByteSource.TERMINATOR : ByteSource.LT_NEXT_COMPONENT;
+                upper = v -> type.asComparableBytes(ByteBufferAccessor.instance, exp.upper.value.raw, v, terminator);
+            }
```

Comment:
I just noticed this doesn't work correctly. We need to fix it before merging to prevent unnecessary false positives for more keys than the query key.